src/cpu/cpu.rs

```diff
@@ -116,7 +116,7 @@ impl Cpu {
                         let rs = self.read_reg_gpr(instr.rs()) as u32;
                         let rt = self.read_reg_gpr(instr.rt()) as u32;
 
-                        let res = ((rs * rt) as i32) as u64;
+                        let res = (rs as u64) * (rt as u64);
 
                         // TODO: Undefined if last 2 instructions were
```

Comment:
I agree.

Just to expand on that: there are two things that could cause UMULTU to be undefined: the presence of meaningful data in the upper 32 bits of the registers being multiplied, or if an MFHI or MFLO instruction were executed recently.  Checking for those undefined behaviors and warning about them might be a good way to detect bugs in the N64 ROM you are emulating or to detect bugs in your simulator.  Detecting bugs in the N64 ROM probably isn't important, but detecting bugs in the simulator would be useful.